### PR TITLE
Storage: remove LFS

### PR DESCRIPTION
**Pull request by skotopes**
# What's new

- Remove the LFS partition for more flash space
- Redirect `/int` requests to SD card (`/ext`)
- Remove the MoveToSD tool
- Load settings automatically on SD card insertion for the following:
  - Bluetooth (settings and keys)
  - Desktop
  - Dolphin
  - Expansion
  - Notification
  - SubGHz region (now separate as RegionSrv)
- Speed up BLE
- Store the PIN code in RTC
- Fix FL-3547

# TODO

- [x] PIN 
- [x] Settings on SD insert
- [x] Startup without SD 
- [x] Various lags when loading files from slow SD
- [x] Testing

# Verification 

`./fbt LIB_DEBUG=1 flash_usb_full` and full set of integration tests

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: targets/f7/furi_hal/furi_hal_region.c [hunk 1]
```diff
@@ -71,66 +71,79 @@ const FuriHalRegion furi_hal_region_jp = {
...
 const FuriHalRegion* furi_hal_region_get(void) {
-    return furi_hal_region;
+    const FuriHalVersionRegion region = furi_hal_version_get_hw_region();
+    const FuriHalRegion* ret;
+
+    if(region < FuriHalVersionRegionWorld) {
+        ret = furi_hal_static_regions[region];
+
+    } else {
+        furi_mutex_acquire(furi_hal_dynamic_region_mutex, FuriWaitForever);
```
Comment (skotopes) [inline targets/f7/furi_hal/furi_hal_region.c]: furi_check furi_mutex_* return results

File: targets/f7/furi_hal/furi_hal_region.c [hunk 2]
```diff
@@ -71,66 +71,79 @@ const FuriHalRegion furi_hal_region_jp = {
...
+    }
+
+    return ret;
 }
 
 void furi_hal_region_set(FuriHalRegion* region) {
     furi_check(region);
 
-    furi_hal_region = region;
+    furi_mutex_acquire(furi_hal_dynamic_region_mutex, FuriWaitForever);
```
Comment (skotopes) [inline targets/f7/furi_hal/furi_hal_region.c]: same

File: furi/core/event_loop.c [hunk 3]
```diff
@@ -57,6 +57,9 @@ FuriEventLoop* furi_event_loop_alloc(void) {
     ulTaskNotifyValueClearIndexed(
         instance->thread_id, FURI_EVENT_LOOP_FLAG_NOTIFY_INDEX, 0xFFFFFFFF);
 
+    furi_thread_set_signal_callback(
```
Comment (skotopes) [inline furi/core/event_loop.c]: see line 180

File: applications/settings/desktop_settings/scenes/desktop_settings_scene_favorite.c [hunk 4]
```diff
@@ -212,21 +212,25 @@ bool desktop_settings_scene_favorite_on_event(void* context, SceneManagerEvent e
                 strncpy(
                     curr_favorite_app->name_or_path,
                     furi_string_get_cstr(temp_path),
-                    MAX_APP_LENGTH);
+                    sizeof(curr_favorite_app->name_or_path));
```
Comment (skotopes) [inline applications/settings/desktop_settings/scenes/desktop_settings_scene_favorite.c:215]: why?

File: applications/services/region/region.c [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        furi_timer_pending_callback(region_loader_pending_callback, furi_thread_get_current(), 0);
+    }
+}
+
+static void region_storage_callback(const void* message, void* context) {
+    UNUSED(context);
+    const StorageEvent* event = message;
+
+    if(event->type == StorageEventTypeCardMount) {
+        FuriThread* loader = furi_thread_alloc_ex(NULL, 2048, region_load_file, NULL);
```
Comment (skotopes) [inline applications/services/region/region.c:128]: sketchy af

File: applications/settings/desktop_settings/scenes/desktop_settings_scene_favorite.c (hunk 4 above)
Comment (gsurkov) [reply applications/settings/desktop_settings/scenes/desktop_settings_scene_favorite.c:215]: MAX_APP_LENGTH might be some weird number (for example, `name_or_path` might be defined as `char name_or_path[MAX_APP_LENGTH + 1]`), better rely on the type information.

File: lib/toolbox/crc32_calc.c [hunk 6]
```diff
@@ -3,9 +3,39 @@
...
+    for(size_t i = 0; i < size; i++) {
+        crc = (crc >> 4) ^ rtable[(crc ^ (data[i] >> 0)) & 0xf];
+        crc = (crc >> 4) ^ rtable[(crc ^ (data[i] >> 4)) & 0xf];
+    }
+
+    return crc;
+}
+
 uint32_t crc32_calc_buffer(uint32_t crc, const void* buffer, size_t size) {
     // TODO FL-3547: consider removing dependency on LFS
```
Comment (hedger) [inline lib/toolbox/crc32_calc.c]: stale todo + why keep lfs in name now?

File: applications/settings/desktop_settings/application.fam [hunk 7]
```diff
@@ -7,6 +7,6 @@ App(
         "desktop",
         "gui",
     ],
-    stack_size=1 * 1024,
+    stack_size=2 * 1024,
```
Comment (hedger) [inline applications/settings/desktop_settings/application.fam]: x2 increase?

File: applications/services/storage/storage_processing.c [hunk 8]
```diff
@@ -60,36 +60,28 @@ static StorageType storage_get_type_by_path(FuriString* path) {
...
-            break;
-        }
-    }
-}
 
 static FS_Error storage_get_data(Storage* app, FuriString* path, StorageData** storage) {
     StorageType type = storage_get_type_by_path(path);
 
     if(storage_type_is_valid(type)) {
+        // Any storage phase-out: redirect "/any" to "/ext"
```
Comment (hedger) [inline applications/services/storage/storage_processing.c:70]: maybe also emit a warning to logs when accessing /any?

File: applications/services/region/region.c [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            free(pb_region.bands.arg);
+            break;
+        }
+
+        FuriHalRegion* region = pb_region.bands.arg;
+
+        memcpy(
+            region->country_code,
+            pb_region.country_code->bytes,
+            pb_region.country_code->size < 4 ? pb_region.country_code->size : 3);
```
Comment (hedger) [inline applications/services/region/region.c]: sizeof() would be cleaner

File: applications/services/desktop/views/desktop_view_pin_input.c [hunk 10]
```diff
@@ -6,14 +6,16 @@
...
-#include <desktop/desktop_settings.h>
 
 #define NO_ACTIVITY_TIMEOUT 15000
 
 #define PIN_CELL_WIDTH 13
 #define DEFAULT_PIN_X 64
 #define DEFAULT_PIN_Y 32
 
+#define MIN_PIN_LENGTH 4
+#define MAX_PIN_LENGTH 10
```
Comment (hedger) [inline applications/services/desktop/views/desktop_view_pin_input.c]: we also have `#define DESKTOP_PIN_CODE_MAX_LEN (14)`

File: applications/services/desktop/helpers/pin_code.c [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    return furi_hal_rtc_get_pin_value() == desktop_pin_code_pack(pin_code);
+}
+
+bool desktop_pin_code_is_equal(const DesktopPinCode* pin_code1, const DesktopPinCode* pin_code2) {
+    furi_check(pin_code1);
+    furi_check(pin_code1->length <= DESKTOP_PIN_CODE_MAX_LEN);
+    furi_check(pin_code2);
+    furi_check(pin_code2->length <= DESKTOP_PIN_CODE_MAX_LEN);
+
+    if(pin_code1->length == pin_code2->length) {
```
Comment (hedger) [inline applications/services/desktop/helpers/pin_code.c]: return (->length == ->length) && (memcmp() == 0)?

File: applications/services/desktop/desktop_settings.c [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        if(version == DESKTOP_SETTINGS_VER) {
+            success = saved_struct_load(
+                DESKTOP_SETTINGS_PATH,
+                settings,
+                sizeof(DesktopSettings),
+                DESKTOP_SETTINGS_MAGIC,
+                DESKTOP_SETTINGS_VER);
+
+        } else if(version == DESKTOP_SETTINGS_VER_10) {
+            DesktopSettingsV10 settings_v10;
```
Comment (hedger) [inline applications/services/desktop/desktop_settings.c]: increasing service's stack size for that thing is rich, maybe malloc()?

File: applications/services/desktop/desktop_i.h [hunk 13]
```diff
@@ -56,42 +55,37 @@ struct Desktop {
...
+    FuriSemaphore* animation_semaphore;
 
     DesktopClock clock;
+    DesktopSettings settings;
 
-    bool in_transition : 1;
-    bool locked : 1;
-
-    FuriSemaphore* animation_semaphore;
+    bool in_transition;
```
Comment (hedger) [inline applications/services/desktop/desktop_i.h:85]: not a big thing, but why no longer packed bitfields?

File: targets/f7/furi_hal/furi_hal_region.c [hunk 14]
```diff
@@ -71,66 +71,80 @@ const FuriHalRegion furi_hal_region_jp = {
...
+    furi_hal_dynamic_region_mutex = furi_mutex_alloc(FuriMutexTypeNormal);
 }
 
 const FuriHalRegion* furi_hal_region_get(void) {
-    return furi_hal_region;
+    const FuriHalVersionRegion region = furi_hal_version_get_hw_region();
+    const FuriHalRegion* ret;
+
+    if(region < FuriHalVersionRegionWorld) {
+        ret = furi_hal_static_regions[region];
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_region.c]: wouldn't that completely prevent applying provisioned region for r00-r03 devices?

File: applications/services/desktop/desktop.c [hunk 15]
```diff
@@ -205,84 +217,57 @@ static void desktop_clock_timer_callback(void* context) {
...
-    } else {
-        view_port_enabled_set(desktop->clock_viewport, false);
-    }
-}
-
-void desktop_lock(Desktop* desktop) {
-    furi_assert(!desktop->locked);
-
-    furi_hal_rtc_set_flag(FuriHalRtcFlagLock);
+    const bool clock_enabled = gui_active_view_port_count(desktop->gui, GuiLayerStatusBarLeft) < 6;
```
Comment (hedger) [inline applications/services/desktop/desktop.c:220]: 🤔

File: applications/services/desktop/desktop.c [hunk 16]
```diff
@@ -205,84 +217,57 @@ static void desktop_clock_timer_callback(void* context) {
...
-        desktop->scene_manager, DesktopSceneLocked, SCENE_LOCKED_FIRST_ENTER);
-    scene_manager_next_scene(desktop->scene_manager, DesktopSceneLocked);
+    view_port_enabled_set(desktop->clock_viewport, clock_enabled);
+}
 
-    DesktopStatus status = {.locked = true};
-    furi_pubsub_publish(desktop->status_pubsub, &status);
+static bool desktop_check_file_flag(const char* flag_path) {
+    Storage* storage = furi_record_open(RECORD_STORAGE);
+    bool exists = storage_common_stat(storage, flag_path, NULL) == FSE_OK;
```
Comment (hedger) [inline applications/services/desktop/desktop.c]: storage_file_exists()
also - maybe store `Storage*` in desktop object?

File: applications/settings/desktop_settings/application.fam (hunk 7 above)
Comment (gsurkov) [reply applications/settings/desktop_settings/application.fam]: Was needed at some point, but can be reverted back safely now.

File: applications/services/desktop/desktop_settings.c (hunk 12 above)
Comment (gsurkov) [reply applications/services/desktop/desktop_settings.c]: That is one big fat struct indeed, better to malloc() it.

File: applications/services/desktop/desktop.c (hunk 16 above)
Comment (gsurkov) [reply applications/services/desktop/desktop.c]: done

File: applications/services/desktop/helpers/pin_code.c (hunk 11 above)
Comment (gsurkov) [reply applications/services/desktop/helpers/pin_code.c]: done

File: targets/f7/furi_hal/furi_hal_region.c (hunk 14 above)
Comment (gsurkov) [reply targets/f7/furi_hal/furi_hal_region.c]: fixed

File: applications/services/desktop/desktop_i.h (hunk 13 above)
Comment (gsurkov) [reply applications/services/desktop/desktop_i.h:85]: Doesn't matter due to mandatory 4 byte padding in the end of the struct

File: lib/toolbox/crc32_calc.c (hunk 6 above)
Comment (gsurkov) [reply lib/toolbox/crc32_calc.c]: Fixed

File: applications/services/desktop/views/desktop_view_pin_input.c (hunk 10 above)
Comment (gsurkov) [reply applications/services/desktop/views/desktop_view_pin_input.c]: Made it 10 everywhere for consistency (but 14 is the threoretical maximum that can be packed into a 32-bit value)

File: applications/services/region/region.c (hunk 9 above)
Comment (gsurkov) [reply applications/services/region/region.c]: Fixed

File: applications/services/storage/storage_processing.c (hunk 8 above)
Comment (gsurkov) [reply applications/services/storage/storage_processing.c:70]: Done

File: applications/services/desktop/desktop.c (hunk 15 above)
Comment (gsurkov) [reply applications/services/desktop/desktop.c:220]: 🤷

File: targets/f7/ble_glue/hw_ipcc.c [hunk 17]
```diff
@@ -113,10 +115,12 @@ void HW_IPCC_SYS_Init(void) {
 void HW_IPCC_SYS_SendCmd(void) {
     LL_C1_IPCC_SetFlag_CHx(IPCC, HW_IPCC_SYSTEM_CMD_RSP_CHANNEL);
 
-    FuriHalCortexTimer timer = furi_hal_cortex_timer_get(33000000);
+    FuriHalCortexTimer timer = furi_hal_cortex_timer_get(IPCC_SEND_CMD_TIMEOUT_US);
 
     while(LL_C1_IPCC_IsActiveFlag_CHx(IPCC, HW_IPCC_SYSTEM_CMD_RSP_CHANNEL)) {
         furi_check(!furi_hal_cortex_timer_is_expired(timer), "HW_IPCC_SYS_SendCmd timeout");
+        // Give time for FreeRTOS timers if kernel is running
+        furi_delay_ms(5);
```
Comment (skotopes) [inline targets/f7/ble_glue/hw_ipcc.c]: Main user of this api is deep sleep routine, sleep shouldn't bee allowed there

File: scripts/testops.py [hunk 18]
```diff
@@ -39,7 +39,9 @@ def _get_flipper(self, retry_count: Optional[int] = 1):
 
             if port := resolve_port(self.logger, self.args.port):
                 self.logger.info(f"Found flipper at {port}")
+                time.sleep(1)
```
Comment (skotopes) [inline scripts/testops.py:42]: why?

File: applications/services/bt/bt_service/bt.c [hunk 19]
```diff
@@ -427,6 +439,91 @@ static void bt_close_connection(Bt* bt, BtMessage* message) {
...
+}
+
+static void bt_handle_reload_keys_settings(Bt* bt) {
+    bt_load_keys(bt);
+    bt_start_application(bt);
+    bt_load_settings(bt);
+}
+
+static void bt_init_keys_settings(Bt* bt) {
+#ifndef STORAGE_INT_ON_LFS
```
Comment (skotopes) [inline applications/services/bt/bt_service/bt.c]: Should we drop STORAGE_INT_ON_LFS flag?

Originally it was proof of concept where I wanted to be able to switch between different builds

File: scripts/testops.py (hunk 18 above)
Comment (gsurkov) [reply scripts/testops.py:42]: Without it, the port is opened too early which results in an unhandled exception

File: targets/f7/ble_glue/hw_ipcc.c (hunk 17 above)
Comment (gsurkov) [reply targets/f7/ble_glue/hw_ipcc.c]: Seems to work fine now without it, magic

**Review by hedger — APPROVED** [review]

**Pull request merged**